File: java/0300_lengthOfLIS.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length];
        dp[0]=1;
        int res=1;
        for(var i=1;i<nums.length;i++)
        {
            dp[i]=1;
            for(var j=0;j<i;j++)
                if(nums[i]>nums[j])
                    dp[i] = Math.max(dp[i],dp[j]+1);
            res = Math.max(res,dp[i]);
        }
        return res;
    }
}

File: cpp/0300_lengthOfLIS.cpp
// dp: 用dp[i]代表以第i个位置为结尾的最长序列长度
// dp[i] = max(dp[argmax_{nums[idx]<nums[i]}]+1,dp[i]),if idx exist; else 1
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(),1);
        int res=1;
        for(auto i=1;i<nums.size();i++)
        {
            for(auto j=0;j<i;j++)
                if(nums[i]>nums[j])
                    dp[i] = max(dp[j]+1,dp[i]);
                
            res = max(dp[i],res);
        }
        return res;
    }
};
